### ISSUE TITLE
KeyError raised when using nested serializers

### ISSUE THREAD

**Issue by thatguysimon** [InterSIS/django-rest-serializer-field-permissions#15]
Hi, first of all, great work on this package!

I'm trying to use it along with DRF's nested serializers feature.
My code looks something like this:

```python
class AccountSerializer(serializers.ModelSerializer):
    email = fields.EmailField(permission_classes=(IsAuthenticated(), ))

    class Meta:
         model = Account

class ContactSerializer(serializers.ModelSerializer):
    account = AccountSerializer(read_only=True)

    class Meta:
        model = Contact
        fields = (..., 'account', )
```

Unfortunately I'm getting the following error when calling the endpoint:
```
File ".../rest_framework_serializer_field_permissions/serializers.py", line 24, in fields
    request = self._context["request"]
KeyError: 'request'
```

Looks like the context is not passed to the nested serializer, so we can't access the request.

I used the workaround suggested here https://github.com/encode/django-rest-framework/issues/2555#issuecomment-253201525 and it works but is there a nicer way to solve this? 
I can help out and open a PR if you have an idea.

**Comment by richardARPANET** [comment]
The linked solution didn't work for me, instead I did this:

1. Install https://pypi.org/project/django-middleware-global-request/
2. Edit `FieldPermissionSerializerMixin`



```python

from django_global_request.middleware import get_request
from rest_framework.exceptions import ValidationError


class FieldPermissionSerializerMixin(object):
    """
    Mixin to your serializer class as follows:

        class PersonSerializer(FieldPermissionSerializerMixin, serializers.ModelSerializer):

            family_names = fields.CharField(permission_classes=(IsAuthenticated(), ))
            given_names = fields.CharField(permission_classes=(IsAuthenticated(), ))
    """

    @property
    def fields(self):
        """
        Supercedes drf's serializers.ModelSerializer's fields property
        :return: a set of permission-scrubbed fields
        """
        ret = super(FieldPermissionSerializerMixin, self).fields
        request = get_request()

        forbidden_field_names = [
            field_name for field_name, field in ret.items() if hasattr(field, 'check_permission') and (not field.check_permission(request))
        ]

        for field_name in forbidden_field_names:
            ret.pop(field_name)

        return ret
```

I changed:

```
request = self._context["request"]
```

to

```
request = get_request()
```

**Comment by GoldenWings** [comment]
@thatguysimon @richardARPANET can you make pull a request for this fix?

**Comment by GoldenWings** [comment]
@JASchilz Can you look into this please?

**Comment by JASchilz** [comment]
Sure! I'll look into it, and have more of a response by this weekend.

**Comment by JASchilz** [comment]
Alright. Using `self._context["request"]` has always been a bit of a hack: by putting that underscore in, the DRF developers are making it clear that they make no promises about that attribute or what it will contain.

I think the way forward is along the lines of Richard's comment: middleware that has the responsibility of grabbing the request and making it available. That suggests two possible approaches:

  1. Require the Django Global Request Middleware.
  2. Make our own that does the same thing.

I'm inclined towards (2). It should have a similar interface to (1), but also it would be important that if somebody tries to get the request from the middleware and the request hasn't been set, then the exception should include a helpful error message indicating that the middleware apparently hasn't been included in `settings.py`.

It's pretty straightforward middleware, if anyone wants to tackle it. If not, I'll get it out next weekend!

**Comment by thatguysimon** [comment]
Hey @JASchilz - any news on this?

**Comment by JASchilz** [comment]
Sure! I've worked towards getting this solved.

The basic problem is that the way that the code was grabbing the `request` was kind of a hack. So I've implemented something along the lines of what Richard suggests above: I've created a middleware class that captures the request.

You can see my changes in https://github.com/InterSIS/django-rest-serializer-field-permissions/tree/grab-request-from-middleware . I still have to test this out in a full Django application, but if you follow the README in that branch you might very well find that it works: I've published an alpha version `2.0a1` that you will have to install specifically in order to make that example work.

The `pip` command to install this specific version of the library is:

```
$: pip install django-rest-serializer-field-permissions==2.0a1
```

Once I've tested it out myself, I'll be able to make a `2.0.0` release. If you do get a chance to try it out, let me know how it works!

**Comment by thatguysimon** [comment]
Hey @JASchilz,
We were using version 2.0.0 and it suddenly disappeared from the releases in PyPI 😢 .
Is it related to your comment yesterday?

**Comment by JASchilz** [comment]
Hi @thatguysimon I can work to fix this quickly.

I made a couple of alpha 2.0 releases that didn't have the version number I wanted, so I deleted them within an hour or so of release. I can't find a `2.0.0` release in my history, though.

1. Could I see your `pip freeze`. ?
2. This release you got from PyPi, did it have the new middleware fix or no?

**Comment by JASchilz** [comment]
Once I know the version number and the features it had, I can try to re-release that version to PyPi. :)

**Comment by thatguysimon** [comment]
@JASchilz We've been using 2.0.0 since July 2019, it's not something new. 
I downgraded the package to 1.0.1 this morning so it's ok now, no worries :)
Looking forward to check out 2.0.0 when it's out. Thanks!

**Comment by JASchilz** [comment]
Alright. Thanks for letting me know about this! Just to make sure I don't cause errors for anyone else, version 2.0.0 is now available again, __without__ the changes discussed above.

Version 3.0.0 will include the new middleware changes discussed above.

**Comment by JASchilz** [comment]
I believe that with #20 and release of 3.0.0 this should now be resolved.

I haven't tested this release with nested serializers specifically, but I have implemented something like Richard's solution into the library, and this solution does address the specific error that you (@thatguysimon) were encountering.

To apply this to your application, check out the [revised tutorial](https://github.com/InterSIS/django-rest-serializer-field-permissions#complete-tutorial). I think you'll need to:
  1. Upgrade your version of the field permissions library: `pip install django-rest-serializer-field-permissions~=3.0`
  2. Then add the new middleware class to your app's `MIDDLEWARE`. See the README/tutorial.

**Comment by thatguysimon** [comment]
@JASchilz - Looks good on our end! 
Thanks! 🎉

**Comment by ajhodges** [comment]
I think this can actually be resolved without using a middleware by using `self.context` instead of `self._context` to retrieve the request. Was there a reason for not using `self.context`? I've added tests (confirmed failing with a KeyError with the old `self._context` approach) for this in a branch on my fork and it's working fine for me. I will open a PR for review.

**Comment by thatguysimon** [comment]
Hey @JASchilz,
I found myself returning to this issue now because I've encountered a problem with the current solution.

When using the middleware to grab the request, the request object that you're grabbing is an instance of `WSGIRequest` from Django core. However, the request from `context` that was used before is an instance of `rest_framework.request.Request` which is a wrapper of the former.

When writing field permission classes, using `request.user` works fine as long as `AuthenticationMiddleware` is performed before `RequestMiddleware`. 

However, if you put anything custom on the request object from anywhere else in DRF, you're going to put it on the instance of `rest_framework.request.Request`, so it will **not be available on the request object** (of type `WSGIRequest`) by the time you want to use it in the field permission class.

So for example if I do this anywhere (for example in a DRF permission class):
```python
request.something_important = "123"
```

Then In a field permission class this will result in an error:
```python
class IsPermitted(BaseFieldPermission):
    def has_permission(self, request):
        return request.something_important == "123"  # => AttributeError!
```


If @ajhodges's solution at #21 indeed fixes the original problem raised in this issue, then I think that's the best solution because a) it's the cleanest and b) it will pass an instance of `rest_framework.request.Request` to the field permission classes as expected.
